### DISCUSSION TITLE
Roadmap

### DISCUSSION THREAD

**Discussion by wjk** [PureDarwin/PureDarwin#99] [Announcements]
This post will be updated to show what tasks and projects that we are currently working on or would appreciate help with, and tasks and projects that are on our longer-term plan for PureDarwin. Community contributions towards implementing anything here are welcome.

# Current Tasks

1. Fix https://github.com/PureDarwin/PureDarwin/pull/95. I am passing the `-bind_at_load` flag to ld just as I should, but inexplicably ld ignores it and emits lazy-stubs anyways, which is an error in libplatform because it is lower level than libdyld. Debugger analysis does not reveal anything more than that it successfully has received the flag, and that it is still emitting lazy symbols.

# Roadmap

1. Get xnu to boot, at least until it tries to start launchd. This requires that a number of kexts be implemented, some of which may need to be written from scratch.
2. Add launchd and launchctl, and make them able to spawn an interactive shell.
3. Implement a terminal-based installer that will format a drive as HFS+ and copy PureDarwin into it.

# Wishlist

1. Read/write APFS support. Everything we will need is [documented in this PDF](https://developer.apple.com/support/downloads/Apple-File-System-Reference.pdf). I would suggest first building a library that can read and write APFS data in abstract data buffers, and then wrap it in a kext so that it can be used on PD. Note that such a library can be developed separately from the rest of the repository, so nothing is stopping anyone from starting on it today.
2. Foundation and Foundation-level model frameworks
3. CoreGraphics (enough to run an arbitrary self-contained GUI)
4. AppKit (and the rest of the public macOS GUI infrastructure)
5. A macOS-like desktop and applications!

**Comment by kjarex** [comment]
Which kexts are missing to get xnu to boot?

**Comment by bliu13** [comment]
Since MacOS 26 is going to be the last version that officially supports Intel-based systems, would it be appealing to use this as the base for PureDarwin?

**Comment by hideout** [comment]
yes ofc we will have a deeper look on MacOS26

and then let’s see what we can do with the future of

PureDarwin and the new arm arch.

take care

Ferdinand

Bart

/-\


Benson Liu ***@***.***> schrieb am Mi. 18. Juni 2025 um 18:33:
